doc/klip.1
.TH KLIP 1 2026-08-02 "0.3.0" "User Commands"
.
.
.SH NAME
klip \- copy/paste anything over the network
.
.
.SH SYNOPSIS
.sp
\fBklip\fP [\fIOPTIONS\fP] <\fICOMMAND\fP>
.sp
\fBklip\fP [\fIOPTIONS\fP] \fBhelp\fP [\fICOMMAND\fP]
.sp
\fBklip\fP [\fIOPTIONS\fP] \fBversion\fP
.sp
\fBklip\fP [\fIOPTIONS\fP] \fB\-\-help\fP
.sp
\fBklip\fP [\fIOPTIONS\fP] \fB\-\-version\fP
.
.
.SH DESCRIPTION
klip is a tool for copying and pasting anything over the network. It works by
sending content to an intermediary server, which can then be retrieved by any
client with the correct credentials. This allows for sharing data across hosts
that are behind different NATs, firewalls, or are otherwise unable to
communicate directly.
.sp
The server does not know much about the data it is handling, and can therefore
be exposed to the public internet without much risk.
.sp
klip reads from standard input and writes to standard output. For the sake of
simplicity and ease of maintenance, there are no options to change the input
or output to point to files or sockets or anything else. Instead, one can use
pipes or redirection to achieve the desired effect.
.
.
.SH COMMANDS
.TP 12
\fIc\fP, \fIcopy\fP
.RS 4
Send the content of standard input to the klip clipboard.
.sp
The clipboard is not persisted anywhere and has no history. This command will
overwrite any existing content.
.RE
.TP 12
\fIp\fP, \fIpaste\fP
.RS 4
Retrieve content from the klip clipboard and write it to standard output.
.sp
The content on the server is left intact until the next \fIcopy\fP or
\fImove\fP command.
.RE
.TP 12
\fIm\fP, \fImove\fP
.RS 4
Retrieve and delete content from the klip clipboard and write it to standard
output.
.sp
This operation is atomic, meaning that the content is deleted only after all
existing clients have finished interacting with it.
.RE
.TP 12
\fIserve\fP [\fIOPTIONS\fP]
.RS 4
Start a klip server.
.sp
This command must be run on a machine that is accessible to all clients that
you wish to share content with. Like other commands, this is also silent and
there are no access logs.
.RE
.TP 12
\fIgenkeys\fP [\fIOPTIONS\fP]
.RS 4
Generates all required keys and writes the resultant configuration to standard
output.
.sp
The configuration printed includes the default listen and connect addresses
and is formatted as a TOML document for ease of copying into a configuration
file.
.sp
Not all keys are actually necessary for both clients and servers. There are
separate sections in the output for each use case, and it is recommended to
only store what is necessary in the configuration file for each host.
.RE
.TP 12
\fIhelp\fP [\fICOMMAND\fP]
.RS 4
Display help information for a specific command.
.sp
If no command is specified, this will display general help information.
.RE
.TP 12
\fIversion\fP
.RS 4
Display version information.
.RE
.
.
.SH OPTIONS
This section documents all flags that klip accepts. Flags are grouped
according to the commands they are associated with.
.
.SS KLIP OPTIONS
.sp
\fI\-c\fP \fIFILE\fP, \fI\-\-config\fP=\fIFILE\fP
.RS 4
Load configuration from \fIFILE\fP. If this flag is not specified, klip will
look for a config file at \fI$HOME/.klip.toml\fP.
.RE
\fI\-h\fP, \fI\-\-help\fP
.RS 4
Display help information for the command.
.RE
\fI\-V\fP, \fI\-\-version\fP
.RS 4
Display version information.
.RE
.
.SS SERVE OPTIONS
.sp
\fI\-\-max\-clients\fP=\fINUM\fP
.RS 4
Set the maximum number of simultaneous client connections. Defaults to 10.
.RE
\fI\-\-max\-len\-mb\fP=\fINUM\fP
.RS 4
Set the maximum content length to accept in MiB. Set to 0 for no limit (which
is the default).
.RE
\fI\-t\fP \fITIMEOUT\fP, \fI\-\-timeout\fP=\fITIMEOUT\fP
.RS 4
Set the connection timeout in seconds. Defaults to 10.
.sp
This refers only to the timeout for establishing a connection. Use the
\fI\-\-data\-timeout\fP flag to set the data transmission timeout.
.RE
\fI\-\-data\-timeout\fP=\fITIMEOUT\fP
.RS 4
Set the data transmission timeout in seconds. Defaults to 3600.
.RE
.
.SS GENKEYS OPTIONS
.sp
\fI\-p\fP, \fI\-\-password\fP
.RS 4
If set, the keys are generated deterministically from a password. The password
may be entered at the subsequent prompt or piped to standard input. Unlike
most other command line tools, the password will be visible while typing.
.sp
This flag is not recommended to be set on most systems, and randomly generated
keys should be used instead. This is provided because it is useful in some
airgapped systems.
.RE
.
.
.SH EXIT STATUS
After successful execution, the exit status of the program is \fB0\fP. If an
error occurred, the exit status is \fB1\fP. If invalid command-line arguments
were provided, the exit status is \fB2\fP.
.sp
.IP \(bu 3n
\fB0\fP exit status indicates that the command was successful.
.
.IP \(bu 3n
\fB1\fP exit status indicates that an error occurred. This is true for both
catastrophic errors and for soft errors.
.
.IP \(bu 3n
\fB2\fP exit status indicates that the command-line arguments were invalid.
.
.
.SH CONFIGURATION FILES
klip reads its configuration from a TOML file. The default location for this
file is \fI$HOME/.klip.toml\fP. This can be overridden with the
\fI\-\-config\fP option.
.sp
A configuration file can be generated using the \fIgenkeys\fP command. Each
invocation of this command produces a set of keys for the server and client.
Clients that share these keys can all share content with each other through
the server.
.sp
Only the \fBlisten\fP and \fBconnect\fP fields are meant to be edited
manually. These fields represent the address that the server should bind to,
and the address that the client should connect to, respectively.
.
.
.SH SHELL COMPLETION
Shell completion files are included in the release tarball for Bash, Fish,
Zsh, PowerShell, and Elvish.
.sp
For \fBbash\fP, move \fBklip.bash\fP to \fB$XDG_CONFIG_HOME/bash_completion\fP
or \fB/etc/bash_completion.d/\fP.
.sp
For \fBfish\fP, move \fBklip.fish\fP to \fB$HOME/.config/fish/completions/\fP.
.sp
For \fBzsh\fP, move \fB_klip\fP to one of your \fB$fpath\fP directories.
.
.
.SH VERSION
0.3.0
.
.
.SH HOMEPAGE
\fIhttps://github.com/lmaotrigine/klip\fP
.sp
Please report bugs and feature requests to the issue tracker. Please do your
best to provide a reproducible test case for bugs.
.
.SH AUTHORS
Isis Ebsen <\fIisis@5ht2.me\fP>


CHANGELOG.md
TBD
===
Unreleased changes. Release notes have not yet been written.

0.3.0
=====
Bug fixes:

* Fix potential panic in clients when the server terminates the connection in
  the middle of a paste or move operation.
* Fix server becoming unresponsive during long-running move operations.

Feature enhancements:

* Exclude --help and --version from tab completion if a subcommand has been
  entered.
* Provide a more informative error message when the server terminates the
  connection.
* Provide visual feedback when typing passwords.
* Support SIGINFO on illumos.

Miscellaneous:

* Return an error when the `connect` or `listen` fields are invalid in the
  config file instead of silently falling back to the default.
* Note in the man page that failure to parse CLI arguments results in exit
  code 2.

0.2.0
=====
Platform support:

* Since the release process is now automated through GitHub Actions, many more
  platforms are now included in the release binaries for klip.

Bug fixes:

* Fix move operations deleting clipboard content despite I/O failure
  [#1](https://github.com/lmaotrigine/klip/pull/1)
* Fix arbitrary timestamps bypassing TTL enforcement
  [#2](https://github.com/lmaotrigine/klip/pull/2)
* Fix build failure when git commit hash cannot be determined.

Miscellaneous:

* Add workflows for CI and release automation.
* Remove enforcement of `lld` for linking.
* Use the RustCrypto ecosystem for cryptography implementations.


completions/klip.bash
#shellcheck disable=SC2207
_klip() {
  local i cur prev opts cmd
  COMPREPLY=()
  if [[ "${BASH_VERSINFO[0]}" -ge 4 ]]; then
    cur="$2"
  else
    cur="${COMP_WORDS[COMP_CWORD]}"
  fi
  prev="$3"
  cmd=""
  opts=""

  for i in "${COMP_WORDS[@]:0:COMP_CWORD}"; do
    case "${cmd},${i}" in
      ",$1")
        cmd="klip"
        ;;
      klip,copy)
        cmd="klip__subcmd__copy"
        ;;
      klip,genkeys)
        cmd="klip__subcmd__genkeys"
        ;;
      klip,help)
        cmd="klip__subcmd__help"
        ;;
      klip,move)
        cmd="klip__subcmd__move"
        ;;
      klip,paste)
        cmd="klip__subcmd__paste"
        ;;
      klip,serve)
        cmd="klip__subcmd__serve"
        ;;
      klip,version)
        cmd="klip__subcmd__version"
        ;;
      klip__subcmd__help,copy)
        cmd="klip__subcmd__help__subcmd__copy"
        ;;
      klip__subcmd__help,genkeys)
        cmd="klip__subcmd__help__subcmd__genkeys"
        ;;
      klip__subcmd__help,help)
        cmd="klip__subcmd__help__subcmd__help"
        ;;
      klip__subcmd__help,move)
        cmd="klip__subcmd__help__subcmd__move"
        ;;
      klip__subcmd__help,paste)
        cmd="klip__subcmd__help__subcmd__paste"
        ;;
      klip__subcmd__help,serve)
        cmd="klip__subcmd__help__subcmd__serve"
        ;;
      klip__subcmd__help,version)
        cmd="klip__subcmd__help__subcmd__version"
        ;;
      *)
        ;;
    esac
  done

  case "${cmd}" in
    klip)
      opts="-c -h -V --config --help --version copy paste move serve genkeys version help"
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        --config)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -c)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__copy)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__genkeys)
      opts="-p --password"
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__help)
      opts="copy paste move serve genkeys version help"
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__help__subcmd__copy)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__help__subcmd__genkeys)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__help__subcmd__help)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__help__subcmd__move)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__help__subcmd__paste)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__help__subcmd__serve)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__help__subcmd__version)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 3 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__move)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
          COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
          return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__paste)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__serve)
      opts="-t -d --max-clients --max-len-mb --timeout --data-timeout"
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
          COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
          return 0
      fi
      case "${prev}" in
        --max-clients)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --max-len-mb)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --timeout)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -t)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        --data-timeout)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        -d)
          COMPREPLY=($(compgen -f "${cur}"))
          return 0
          ;;
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
    klip__subcmd__version)
      opts=""
      if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
        COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
        return 0
      fi
      case "${prev}" in
        *)
          COMPREPLY=()
          ;;
      esac
      COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
      return 0
      ;;
  esac
}

if [[ "${BASH_VERSINFO[0]}" -eq 4 && "${BASH_VERSINFO[1]}" -ge 4 || "${BASH_VERSINFO[0]}" -gt 4 ]]; then
  complete -F _klip -o nosort -o bashdefault -o default klip
else
  complete -F _klip -o bashdefault -o default klip
fi
